ember-server/src/main/scala/org/http4s/ember/server/internal/WebSocketHelpers.scala

```diff
@@ -0,0 +1,213 @@
+/*
+ * Copyright 2019 http4s.org
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.http4s.ember.server.internal
+
+import cats.effect.{Concurrent, Sync}
+import cats.syntax.all._
+import cats.data.NonEmptyList
+import fs2.{Chunk, Pipe, Pull, Stream}
+import fs2.io.tcp._
+import org.http4s.syntax.all._
+import org.http4s._
+import org.http4s.websocket.{FrameTranscoder, WebSocketContext}
+import org.http4s.headers._
+import org.http4s.ember.core.Read
+import org.http4s.ember.core.Util.durationToFinite
+
+import scala.concurrent.duration.Duration
+import java.security.MessageDigest
+import java.util.Base64
+import java.nio.charset.StandardCharsets
+import org.http4s.headers.Connection
+import org.http4s.websocket.WebSocketFrame
+import java.nio.ByteBuffer
+
+object WebSocketHelpers {
```

Comment:
I really want to make this a `WebSocketHandler` class that's parameterized over `F`. I think there's a lot more opportunity to preallocate values, and it's also easier to read since there are less arguments being passed around functions

Reply:
We could do something similar with `ServerHelpers` as well -- a `ServerHandler` that's allocated once per `F`